### ISSUE TITLE
fitz.Point() appears to be on a different grid than rect

### ISSUE THREAD

**Issue by mikeymop** [pymupdf/PyMuPDF#309]
I am stamping a pdf, and trying to enter text just below it.

Do something like so:
```python
srect = (float(coordinates[0]), float(coordinates[1]), float(coordinates[2]), float(coordinates[3]))
page.insertImage(srect, filename=wmark, keep_proportion=True, overlay=True)
stamp_fontsize = int(stamp_text_size) # 12 is most common

t_coor = fitz.Point(float(srect[0] + 2.0) , float(srect[3])) # add top margin to stamp text
print("Text at: ", end="")
print(t_coor)
page.insertText(t_coor, stamp_text, fontsize=stamp_fontsize)
```
I'm have played with different permutations of 
```
t_coor = fitz.Point(float(srect[0] + 2.0) , float(srect[3])) # add top margin to stamp text
```
However it will appear in the same place until I subtract 50 from srect[3]. It as if it's trying to snap to a mediabox below it instead of just taking the absolute coordinates. This means if I stamp somewhere else on another pdf the text might *actually* be at rect[3] - 50 and end up over the stamp itself instead of just 2-4px below the stamp.

**Comment by JorjMcKie** [comment]
this skript:
```python
>>> import fitz
>>> doc=fitz.open("v110-changes.pdf")#just some pdf
>>> page=doc[0]
>>> r=fitz.Rect(100,100,200,200)
>>> page.insertImage(r,filename="nur-ruhig.jpg", keep_proportion=False)
>>> page.insertText(r.bl + (0,20), "hello again!")#text 20 pts below image
1
>>> doc.save("marked.pdf")
>>> 
```
works for me like this:
![grafik](https://user-images.githubusercontent.com/8290722/58833945-29ceb980-8620-11e9-9f2d-e398c055aee9.png)

**Comment by mikeymop** [comment]
So I used the `rect.bl` option you showed me.

Here is the result:
![image](https://user-images.githubusercontent.com/4002969/58834812-466bf100-8622-11e9-8f70-25e3df72d289.png)

```python
srect = fitz.Rect(float(coordinates[0]), float(coordinates[1]), float(coordinates[2]), float(coordinates[3]))
page.insertImage(srect, filename=wmark, keep_proportion=True, overlay=True)

if(stamp_text != False):
    print("Adding text below stamp")
    add_text_below_stamp(page, stamp_text, srect, stamp_text_size)

def add_text_below_stamp(page, stamp_text, srect, stamp_text_size):
    t_coor = srect.bl
    stamp_fontsize = int(stamp_text_size) # 12 is most common
    print("Text at: ", end="")
    print(t_coor)
    page.insertText(t_coor, stamp_text, fontsize=stamp_fontsize, morph=None)
```

**Comment by JorjMcKie** [comment]
Hard to recognize what your expected result actually is.
Coordinates in (Py-) MuPDF are absolute, never relative.
Also I don't understand why you are using those cumbersome ``float()`` functions in the rect definition. You could simply do ``fitz.Rect(coordinates)`` if "coordinates" is some numeric sequence.
Then, if you say "keep_proportion=True", the image will not be distorted, so bottom-left of **visible image** and of the **rect** may not be the same, making the text appear to be misplaced to the left.

Your print info seems to indicate however, that your PDF is one of those malformed ones, which change the page geometry without localizing that - see [here](https://pymupdf.readthedocs.io/en/latest/faq/#misplaced-item-insertions-on-pdf-pages) for a recently updated description of how to handle this.

Basically you need wrap the existing stuff of the page definition in an stack/unstack pair of commands before you insert your own stuff.

**Comment by mikeymop** [comment]
> Hard to recognize what your expected result actually is.

The example you showed me is exactly what I want, I would just set different margins based on the font size, so that if the font size is made large the text would move down further (since the point refers to the bottom left of the first character).

> Coordinates in (Py-) MuPDF are absolute, never relative.
I thought so. Thank you for confirming.

So I heeded your advise and tried sanitizing the pdf using the link you showed me.

First, I installed mutool on ubuntu 19.04, used `mutool -s` to clean the pdf and send that into the program, the text was still far too the left in the same manner as the above picture.

Second, I tried `page._cleanContents()` which crashed on the larger pdf I had pictured because my OS saw the process was using too much memory.

Finally, I worked to implement the best case (option 3) using the following snippet:

```python
for page in source_pdf:
        clean = True
        
        print("Cleaning the pdf contents")
        xreflist = page._getContents()
        print(xreflist)
        c = source_pdf._getXrefStream(xreflist[0])
        source_pdf._updateStream(xreflist[0], b"q\n" + c)
        c = source_pdf._getXrefStream(xreflist[-1])
        source_pdf._updateStream(xreflist[-1], c+b"\nQ")

        page.insertImage(srect, filename=wmark, keep_proportion=True, overlay=True)
```
Before I do any insertions to the pdf.
I declared the `srect` on two different occasions, once manually mapping the coordinates (I pick them up as a json list) and once by casting them to a `tuple()`, they both gave the same value.
`srect = fitz.Rect(coordinates)`

I have also set `keep_proportion=False` so as to remove any discrepancy with `srect` not matching what is actually visibly displayed on the pdf.

Finally, I updated my function for adding the text below the image so that it matches the example you showed me and also prints where it is writing on the coordinate grid.

```python
def add_text_below_stamp(page, stamp_text, srect, stamp_text_size):
    stamp_fontsize = int(stamp_text_size) # 12 is most common
    # l_height = stamp_fontsize * 1.5
    # t_coor = fitz.Point(float(srect[0] + 2.0) , float(srect[3])) # add top margin to stamp text
    print("srect:", end="")
    print(srect)
    print("Text at: ", end="")
    print(srect.bl)
    page.insertText(srect.bl + (0,2), stamp_text, fontsize=stamp_fontsize)
```


I then tried this with two different pdfs. One landscape and one vertical.

pdf 1:
I get the same text distance whether I use 
`page.insertText(srect.bl + (0,2), stamp_text, fontsize=stamp_fontsize)`
or
`page.insertText(srect.bl + (0,5), stamp_text, fontsize=stamp_fontsize)`
until I get well into the negative numbers.
![image](https://user-images.githubusercontent.com/4002969/58911330-9790e900-86e5-11e9-9a07-5ddb349cb870.png)

and on pdf2:

![image](https://user-images.githubusercontent.com/4002969/58911892-e2f7c700-86e6-11e9-8035-7d1e9eaee634.png)

Additionally, I printed out the srect coordinates, and the point coordinates used for the text insertion with each page interation in this script.

```
srect:Rect(2968.0, 2104.0, 3024.0, 2160.0)
Text at: Point(2968.0, 2160.0)
```
So I'm getting the same coordinates as `srect.bl` now however I noticed the coordinates in python aren't the same that I called. For example, on the first pdf pictured above I provided the following

`"108.5","081.7","287.2","392.0"`
Yet picked up the following coordinates:
```
srect:Rect(556.0, 736.0, 612.0, 792.0)
Text at: Point(556.0, 792.0)
```
I'm guessing this is the result of cleaning the pdf using the stack method (option 3).

Do I have to clean the pdf between insertions? I insert an image, a string of text, and a qrcode.png all in one iteration of a for loop.

**Comment by JorjMcKie** [comment]
this looks better but still very weird.
Can you share the PDF so I can try myself?

> Do I have to clean the pdf between insertions? I insert an image, a string of text, and a qrcode.png all in one iteration of a for loop.

You only do this "cleaning" once per page.  After the ``Q`` command, the standard geometry is re-established, so any further text or image insertion should work as expected.
BTW: are your coordinates strings? Then of course they must be converted to numbers to create a Rect ...

**Comment by JorjMcKie** [comment]
When you insert that text, the point ``rect.bl + (0, n)`` has the x-coordinate of the rectangle and a y-coordinate n points below the bottom of the rect. If I understand your intention correctly, tThis value n should equal the height of the text line to insert. So best set it to values between ``n = fontsize * 1.2`` and ``n = fontsize * 1.4``.

**Comment by JorjMcKie** [comment]
@mikeymop  - any update on this? Questions?

**Comment by Randallzoeng** [comment]
Hi @mikeymop , you can pass one more param that `overlay=False`
my scenario kinda like you:
`page.draw_quad(fitz.Quad(r.top_left,r.bottom_left,r.top_right,r.bottom_right),color=(211/255,17/255,69/255),overlay=False)`
`page.draw_quad(fitz.Quad(r.top_left,r.bottom_left,r.top_right,r.bottom_right),color=(211/255,17/255,69/255))`
draw two quad in different place, set overlay=False will be the one you need.

**Comment by mikeymop** [comment]
Hello, my apologies JorjMcKie, I have since transitioned to a new position and this task has come and gone.

The task was ultimately resolved by using trigonometry to perform coordinate translation against a pdf's rotation. This seemed to resolve issues with placing elements when the pdf was rotated visually but not programmatically.

Unfortunately I dont own the code so I cannot share a snippet however I used this [video](https://www.khanacademy.org/computing/pixar/sets/rotation/v/sets-8) as inspiration for the solution.
